include/ghex/communication_object_ipr.hpp

```diff
@@ -0,0 +1,564 @@
+                    void wait() { if (m_wait_fct) m_wait_fct(); }
+
+            };
+
+            /** @brief communication object responsible for exchanging halo data.
+             * Allocates storage depending on the device type and device id of involved fields.
+             * @tparam Communicator communicator type
+             * @tparam GridType grid tag type
+             * @tparam DomainIdType domain id type*/
+            template<typename Communicator, typename GridType, typename DomainIdType>
+            class communication_object_ipr {
+
+                private:
+
+                    // friend class // TO DO: is friend needed?
+                    friend class communication_handle_ipr<Communicator,GridType,DomainIdType>;
+
+                public:
+
+                    // member types
+
+                    /** @brief handle type returned by exhange operation */
+                    using communicator_type = Communicator;
+                    using grid_type = GridType;
+                    using domain_id_type = DomainIdType;
+                    using handle_type = communication_handle_ipr<communicator_type,grid_type,domain_id_type>;
+                    using pattern_type = pattern<communicator_type,grid_type,domain_id_type>;
+                    using pattern_container_type = pattern_container<communicator_type,grid_type,domain_id_type>;
+
+                    template<typename D, typename F>
+                    using buffer_info_type = buffer_info<pattern_type,D,F>;
+
+                private:
+
+                    // member types
+
+                    using address_type = typename communicator_type::address_type;
+                    using index_container_type = typename pattern_type::index_container_type;
+                    using pack_function_type = std::function<void(void*, const index_container_type&, void*)>; // last argument intended as CUDA stream
+
+                    /** @brief pair of domain ids + tag id with ordering */
+                    struct domain_id_pair_and_tag {
+
+                            domain_id_type first_id;
+                            domain_id_type second_id;
+                            int tag;
+
+                            bool operator<(const domain_id_pair_and_tag& other) const noexcept {
+                                return (first_id < other.first_id ? true :
+                                    (first_id > other.first_id ? false :
+                                        (second_id < other.second_id ? true :
+                                            (second_id > other.second_id ? false :
+                                                (tag < other.tag)))));
+                            }
+
+                    };
+
+                    /** @brief Holds a pointer to a set of iteration spaces and a callback function pointer
+                     * which is used to store a field's pack or unpack member function.
```

Comment:
are these ones used for unpack? I guess not.

Reply:
This struct is actually used for the receive part as well, using `std::function<void()>` as function type. I first thought of not using a func at all, but this requires at least to define a new struct (or a specialization of this one), and I think that keeping it as it is is less confusing in the end. I'll mention this in one of the two opened issues anyway.

Reply:
I changed the code a little bit here now. `field_info` is actually used only for the send part, so only with pack function. I added a specific struct for the receive, without callback (unpack) function.